### ISSUE TITLE
Can we please add proper logging for the application.

### ISSUE THREAD

**Issue by PaulMCEscher** [ytdl-org/youtube-dl#31466]
Can we please add proper logging for the application.
for example youtube-dl.exe -l C:\logs\youtube-gl.log --log-level debug
allow for log-levels like info, warning, error, debug
and use a standard log format like
01/13/2023 06:05:00.123 ERROR component=downloading url="https://site.com/xyz/" status="Room is currently offline"
etc.
Redirecting stdout and stderr is from the 80s.  :-)

I'd be open to helping out.
Thanks!

**Comment by dirkf** [comment]
tldr; possibly, if someone can work out how and implement it (suggestion below).

Your proposal reflects typical behaviour for server applications but comparable client applications don't agree on the UI: _wget_ and _curl_ don't have a log-level system, while _ffmpeg_ and _aria2c_ do. Sometimes the '80s were good enough (eg, Heaven17 "Temptation").

Still, if someone else is doing the work, I'd be pleased to consider a PR for this.

As the Manual explains, there is an internal parameter `logger` that allows you to specify a Python logging module instead of writing to `sys.stderr`, etc.

From that you can see that there is already a multi-level diagnostic system:
1. messages written with `to_stdout()`, which go to `sys.stdout` unless the internal flag parameter `logtostderr` is set, and which have DEBUG severity when logged
2. informational messages written with `to_screen(), which are similar to the above but can be muted with `--quiet` 
3. messages through the `report_warning()` method, which go to `sys.stderr` and which have WARNING severity when logged
4. messages written with `to_stderr()`, which go to `sys.stderr` and which have ERROR severity when logged.

See in particular `YoutubeDL.py`, `extractors/common.py`, `utils.py`.

yt-dlp adds:
5. messages written with the `write_debug()` method, which are like `to_stderr()` messages but have DEBUG severity when logged and are muted unless `--verbose` is set.

If making changes in this area, I would support first back-porting the `write_debug()` methods of `YoutubeDL` and `InfoExtractor` from yt-dlp, as above.

Your `--loglevel ...` and `--log/-l ...` options could be implemented with a logging module that is enabled if the latter option is specified. Suppose that a NOTICE severity level between INFO and ERROR is defined, with a corresponding `notice()` method; this could be done with a `LoggingAdapter` that is used with the new file output `Logger` and also when a custom `logger` is specified for an embedded yt-dl instance.

Then the above cases could become, when the logging module is enabled:
1. -> `logger.notice()`
2. -> `logger.info()`
3. -> `logger.warning()`
4. -> `logger.error()`
5. -> `logger.debug()`

By varying `--loglevel ...` one could suppress messages with severity between DEBUG and any of DEBUG, INFO, NOTICE, WARNING and ERROR, with default DEBUG..INFO.

It would be a reasonable guess that
a) quite a few messages of types 1 and 2 ought to be type 5
b) quite a few messages of all types might be of the wrong type.

I think that is probably the major impediment to getting a useful result from this proposal.

Finally, it might be considered to deprecate the existing `to_s...()` APIs and introduce a new API that maps directly to the log levels.

**Comment by PaulMCEscher** [comment]
@dirkf 
Actually I think that I found a workaround using --print-json and redirecting stdout to a log file.

I was interested in finding out which feeds had not been downloaded from in the last 30 days, so I could replace them with new feeds.  The json appears to only be written to when a download starts, so I can figure out the rest.  I'll import the output into a log aggregator, like Splunk, and do some post processing with the data.

I really appreciate your reply.

Thank you!

**Comment by dirkf** [comment]
You can also try the _jq_ utility if you need to process JSON in scripts (ie, not .py scripts).

**Comment by pukkandan** [comment]
Note that are working on completely rewriting the logging in yt-dlp https://github.com/yt-dlp/yt-dlp/pull/5680 though it may be difficult to backport

**Comment by Grub4K** [comment]
Everything in that can be done easily. Namespace meta class magic and enums can just be a namespace. Logging module call stack introspection will have to be completely rewritten or scrapped by replacing the usage of `to_...` completely and using the `Logger` directly. Error stacks and printing will have to be adjusted back to previous state. Other than that I cant see obvious problems with it. I dont know Py2 or codebase well though. Open to help with porting or questions, please ping if that happens.
